pkg/auth/providers/saml/saml_client.go

```
@@ -0,0 +1,385 @@
+package saml
+
+import (
+	"bufio"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/crewjam/saml"
+	"github.com/crewjam/saml/logger"
+	"github.com/crewjam/saml/samlsp"
+	"github.com/gorilla/mux"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+//Client implements a client for the saml library
+type Client struct {
+	config *v3.SamlConfig
+	SamlSP *saml.ServiceProvider
+}
+
+type IDPMetadata struct {
+	XMLName           xml.Name                `xml:"urn:oasis:names:tc:SAML:2.0:metadata EntityDescriptor"`
+	ValidUntil        time.Time               `xml:"validUntil,attr"`
+	EntityID          string                  `xml:"entityID,attr"`
+	IDPSSODescriptors []saml.IDPSSODescriptor `xml:"IDPSSODescriptor"`
+	SPSSODescriptors  []saml.SPSSODescriptor  `xml:"SPSSODescriptor"`
+}
+
+const (
+	redirectBackBase = "redirectBackBase"
+	redirectBackPath = "redirectBackPath"
+)
+
+var Root *mux.Router
+
+func InitializeSamlServiceProvider(configToSet *v3.SamlConfig, name string) error {
+
+	var idpURL string
+	var privKey *rsa.PrivateKey
+	var cert *x509.Certificate
+	var err error
+	var ok bool
+
+	if configToSet.IDPMetadataURL == "" {
+		idpURL = ""
+		if configToSet.IDPMetadataContent == "" {
+			if configToSet.IDPMetadataFilePath == "" {
+				log.Debugf("SAML: Cannot initialize saml SP properly, missing IDP URL/metadata in the config %v", configToSet)
```

should we return an error here?

---

same for lines 73 and/or 78